labrad/decorators.py

```diff
@@ -21,202 +21,140 @@
-                self.accepts, self.returns, self.notes)
+class Setting(object):
+    """
+    Represent a single setting within a Server object.
 
-    def handleRequest(self, c, data):
-        return self.func(c, data)
+    args:
 
-def setting(lr_ID, lr_name=None, returns=[], lr_num_params=2, **params):
-    """Mark a server method as a remotely-accessible setting.
+    func:       Function to implement the setting.
+    lr_ID:      The ID number of the setting used in the labrad protocol
+    lr_name:    The setting name.  By default, this is derived from the function name
+    returns:    The possible labrad type strings returned by the setting
 
-    The only required parameter is an integer ID.  You may
-    also provide a name to override the name of the decorated
-    function.  In addition, accepted types for each of the setting
-    parameters may be provided as named parameters with a list of
-    strings of allowed types.
+    **kw:       Additional keyword arguments indicate arguments to the setting.  Each keyword
+                should be associated with one or more labrad type strings indicating data types
+                it accepts.
     """
-    def decorated(f):
-        args, varargs, varkw, defaults = getargspec(f)
-        args = args[lr_num_params:]
-
-        # handle generators as inlineCallbacks
-        if _isGenerator(f):
-            f = inlineCallbacks(f)
-
-        # make sure that defined params are actually accepted by the function.
-        # having extra params would not affect the running, but it is
-        # unnecessary and hence may indicate other problems with the code
-        for p in params:
+    def __init__(self, func, lr_ID, lr_name, returns, unflatten, **params):
+        self.func = func
+        self.ID = lr_ID
+        self.name = lr_name or func.__name__
+        self.returns = [returns] if isinstance(returns, basestring) else returns
+        self.isSetting = True
+        self.unflatten = unflatten
+        self.description, self.notes = util.parseSettingDoc(func.__doc__)
+
+        ###
+        # We need to validate the arguments.  Things we need to checks:
+        #
+        # 1) All parameters must match function arguments
+        # 2) Function arguments with no specified parameter default to '?'
+        # 3) The empty tag '' is only allowd on the first argument, and only
+        #    if all other arguments are optional.
+        # 4) If more than one argument are required, we expect to always receive
+        #    a tuple and unpack it
+        # 5) If only a single argument is allowed, we never unpack tuples
+        # 6) If both =1 and >1 arguments are allowed, it is ambiguous whether to
+        #    unpack tuples, so this case is not allowed:  The first argument
+        #    cannot be a tuple or '?' tag if the second argument is optional.
+        
+        argspec = inspect.getargspec(self.func)
+        args = argspec.args[2:]
```

Comment:
A comment about where the 2 comes from would be helpful (first two args are self and context object).